jpegenc/subblocks/dct_2d/dct_2d.py

```diff
@@ -0,0 +1,254 @@
+import numpy as np
+from math import sqrt, pi, cos
+import myhdl
+from myhdl import Signal, ResetSignal, intbv, always_comb, always_seq, instance
+from myhdl.conversion import analyze
+
+from jpegenc.subblocks.common import (input_1d_1st_stage, input_interface,
+                                      output_interface, outputs_2d,
+                                      input_1d_2nd_stage)
+from jpegenc.subblocks.dct_1d import dct_1d
+
+
+class dct_2d_transformation(object):
+
+    """2D-DCT Transformation Class
+
+    It is used as a software reference for the 2D-DCT
+    Transformation
+    """
+
+    def __init__(self):
+        """Initialize the DCT coefficient matrix"""
+        const = sqrt(2.0 / 8)
+        a = const * cos(pi / 4)
+        b = const * cos(pi / 16)
+        c = const * cos(pi / 8)
+        d = const * cos(3 * pi / 16)
+        e = const * cos(5 * pi / 16)
+        f = const * cos(3 * pi / 8)
+        g = const * cos(7 * pi / 16)
+
+        self.coeff_matrix = [[a, a, a, a, a, a, a, a],
+                             [b, d, e, g, -g, -e, -d, -b],
+                             [c, f, -f, -c, -c, -f, f, c],
+                             [d, -g, -b, -e, e, b, g, -d],
+                             [a, -a, -a, a, a, -a, -a, a],
+                             [e, -b, g, d, -d, -g, b, -e],
+                             [f, -c, c, -f, -f, c, -c, f],
+                             [g, -e, d, -b, b, -d, e, -g]]
+
+    def dct_2d_transformation(self, block):
+        """2D-DCT software reference"""
+        coeff_matrix = np.asarray(self.coeff_matrix)
+        block = np.asarray(block)
+        block = block - 128
+        # first 1d-dct with rows
+        dct_result = np.dot(coeff_matrix, np.transpose(block))
+        # second 1d-dct with columns
+        dct_result = np.rint(dct_result)
+        dct_result = np.dot(coeff_matrix, np.transpose(dct_result))
+        dct_result = np.rint(dct_result).astype(int)
+        dct_result = dct_result.tolist()
+        return dct_result
+
+
+@myhdl.block
+def dct_2d(inputs, outputs, clock, reset, num_fractional_bits=14, out_prec=10):
```

Comment:
@mkatsimpris can you review [this example](https://gist.github.com/cfelton/595fdf0ca1b8c61264397e37da6d938f) and let me know if a generic interface, like the one in the linked example, will work for your inputs and outputs.

If so, then we should investigate if we can encapsulate the 2D blocks even more, maybe doing [something like this](https://gist.github.com/cfelton/f6a1a1e0cfb2204cd2ee) and [modify this](https://github.com/cfelton/test_jpeg/blob/master/jpegenc/models/buffers/block_buffer.py)


The goal is to make it as generic, modular and readable as possible.